File: Kymdan/src/main/java/com/kymdan/backend/services/khuyen_mai/KhuyenMaiService.java
package com.kymdan.backend.services.khuyen_mai;

import com.kymdan.backend.entity.ChiTietKhuyenMai;
import com.kymdan.backend.entity.KhuyenMai;
import com.kymdan.backend.model.KhuyenMaiDTO;
import com.kymdan.backend.model.ThongBaoDTO;

import java.util.List;

public interface KhuyenMaiService {
    List<KhuyenMai> xemTatCa();

    ThongBaoDTO taoMoi(KhuyenMaiDTO khuyenMaiDTO);

    ThongBaoDTO sua(KhuyenMaiDTO khuyenMaiDTO);

    ThongBaoDTO xoa(String ma);

    KhuyenMai timBangMa(String ma);

    KhuyenMai timBangTen(String ten);

    List<ChiTietKhuyenMai> timChiTietBangMaKhuyenMai(String maKhuyenMai);

    ThongBaoDTO themSanPhamKhuyenMai(String maKhuyenMai, String maSanPham, Integer giamGia);

    ThongBaoDTO suaSanPhamKhuyenMai(Integer maChiTiet, Integer giamGia);

    ThongBaoDTO xoaSanPhamKhuyenMai(Integer maChiTiet);
}

File: Kymdan_front_end/src/app/service/khuyen-mai.service.ts
import {Injectable} from '@angular/core';
import {Observable} from 'rxjs';
import {HttpClient} from '@angular/common/http';

@Injectable({
  providedIn: 'root'
})
export class KhuyenMaiService {
  public readonly API: string = 'http://localhost:8080/khuyenMai';

  constructor(
    public http: HttpClient
  ) {
  }

  xemTatCa(): Observable<any> {
    return this.http.get(this.API + '/xemTatCa');
  }

  timBangMa(ma): Observable<any> {
    return this.http.get(this.API + '/timBangMa/' + ma);
  }

  timChiTietBangMaKhuyenMai(maKhuyenMai): Observable<any> {
    return this.http.get(this.API + '/timChiTietBangMaKhuyenMai/' + maKhuyenMai);
  }

  taoMoi(thongTin): Observable<any> {
    return this.http.post(this.API + '/taoMoi', thongTin, {headers: {skip: 'true'}});
  }

  sua(thongTin): Observable<any> {
    return this.http.post(this.API + '/sua', thongTin, {headers: {skip: 'true'}});
  }

  xoa(ma): Observable<any> {
    return this.http.get(this.API + '/xoa/' + ma);
  }

  themSanPhamKhuyenMai(maKhuyenMai, maSanPham, giamGia): Observable<any> {
    return this.http.get(this.API + '/themSanPhamKhuyenMai/' + maKhuyenMai + '/' + maSanPham + '/' + giamGia);
  }
}
